docs/server datastorage.md

```diff
@@ -0,0 +1,115 @@
+
+Let's look at a few examples:
+
+### EnergyLink
+EnergyLink lets you share you excess energy across multiple clients and games. It uses a team-specific key of `EnergyLink{team}`. The energy value is a numeric value and it's important that games can add and take energy from it in a thread safe way. A little warning, over the course of a long game, many clients can contribute to the energy value making it larger then an int64, while python and json have no issues with this, some programming languages might need to take extra care.
+
+Adding is easy, just a `Set`-`add` operation, for example to add 20:
+```json
+{
+    "cmd": "Set",
+    "key": "EnergyLink0",
+    "default": 0,
+    "operations": [
+        {"operation": "add", "value": 20},
+    ]
+}
+```
+Depleting, however is more complicated; in this scenario we use both the fact that multiple operations are atomic as well as tagging the Set. First we subtract our value 50, then we set the value back to 0 if it went below 0. Lastly, we will look for the corresponding `SetReply` command with the tag we specified, inside the `SetReply` we compare the `original_value` to the `value` to know how much energy we were actually able to subtract.
+
+```json
+{
+    "cmd": "Set",
+    "key": "EnergyLink0",
+    "tag": "7cc04194-b491-4cba-a89e-ef754502f3ff",
+    "default": 0,
+    "want_reply": true,
+    "operations": [
+        {"operation": "add", "value": -50},
+        {"operation": "max", "value": 0},
+    ]
+}
+```
+Response:
+```json
+{
+    "cmd": "SetReply",
+    "key": "EnergyLink0",
+    "tag": "7cc04194-b491-4cba-a89e-ef754502f3ff",
+    "default": 0,
+    "slot": 5,
+    "want_reply": true,
+    "operations": [
+        {"operation": "add", "value": -50},
+        {"operation": "max", "value": 0},
+    ],
+    "original_value": 20,
+    "value": 0
+
+}
+```
+So we only were able to subtract `original_value` - `value` = 20 energy.
+
+### Gifting
+Gifting allows players to gift items they don't need to other games. For example, if you get a health potion but you are at no risk of dying, someone else might have more use for it. Gifting works by slot-specific giftboxes, any client can add a gift to the giftbox, and only the client of the slot the giftbox belongs to is supposed to remove gifts from it. Now it might seem easy with an array of gift objects however while appending to an array is thread safe, removing entries from an array cannot be done thread safe. Therefore, giftboxes are implemented as dictionaries where each gift has its own unique key. New gifts can safely be added using the `update` operation and removal of specific gifts can be done using their unique keys with the `pop` operation. For more information visit the [gifting api documentation](https://github.com/agilbert1412/Archipelago.Gifting.Net/blob/main/Documentation/Gifting%20API.md).

```

Comment:
I'm not sure exactly how I'd want to phrase this but it's probably worth some sort of mention that the gifting docs should be treated as the authoritative source if the content here is in conflict (ie, out of date)